plugin/storage/memory/memory.go

```diff
@@ -170,11 +170,37 @@ func (m *Store) GetTrace(ctx context.Context, traceID model.TraceID) (*model.Tra
 // Spans may still be added to traces after they are returned to user code, so make copies.
 func (m *Store) copyTrace(trace *model.Trace) *model.Trace {
 	return &model.Trace{
-		Spans:    append([]*model.Span(nil), trace.Spans...),
+		Spans:    m.copySpans(trace),
 		Warnings: append([]string(nil), trace.Warnings...),
 	}
 }
 
+func (m *Store) copySpans(trace *model.Trace) []*model.Span {
+	spans := make([]*model.Span, len(trace.Spans))
+	for i := range trace.Spans {
+		spans[i] = m.copySpan(trace.Spans[i])
+	}
+	return spans
+}
+
+// Copy static span attributes and discard the ones that can be altered outside of the store.
+func (m *Store) copySpan(span *model.Span) *model.Span {
+	return &model.Span{
```

Comment:
This isn't really a deep copy, we could run into similar issues if say some enrichers add to logs or tags. It would be safer to marshal to proto and back for a fully independent clone

Reply:
Should we do a deep copy of `model.Trace` or just individual `model.Span` members?

Reply:
prob easier to clone Trace, it's also in proto

Reply:
Amended to do a full Trace clone. Some test values needed to be massaged to make `reflect.DeepEqual` happy.